README
R script to make an "average" track of several GPS tracks by combining/merging them into a "line of best fit". Made by Michiel Faber <http://www.openstreetmap.org/user/Michiel%20Faber> and wrapped and "packaged" by Guttorm Flatabø. This isn’t the perfect way to do it, but it works.

The "test-tracks" directory contains six tracks for the same route that you can test the software with. "averageTrack_reference.gpx" is an average track made with this script.

TopoFusion software uses a more sophisticated method and have written a paper about it <http://www.topofusion.com/network.php>. The issue has also been discussed on the gpsbabel-misc mailing list <http://goo.gl/HJE5s>.

See http://wiki.openstreetmap.org/wiki/Average_tracks for usage instructions.

Home: https://github.com/dittaeva/Average-tracks

This software is dedicated to the public domain through CC0:
http://creativecommons.org/publicdomain/zero/1.0/


changelog.txt
Changelog
Creator: Michiel Faber
Original start date: back in 2009?
script/program: AverageTrack
Description: This program tries to make an 'average' track out of several gps recordings. 
Goal: Ultimately, it should create a network of tracks, similar to the Topofusion software (http://www.topofusion.com).
      For more details read their paper at: http://www.topofusion.com/jcdl-04-trails.pdf.

Versioning for now works Ubuntu style: yearmonthday
------------
# Version:
xxxxxxxx

# Changes:
-

------------
# Version:
20120121

# Changes:
- Output from fLoadFiles() expanded to also outputs the filenames of the imported files.
- Moved several function from single files and "functions.R" to "functionsAverage.R" and "functionsGlobal.R".
- Moved code from fFindTrackWithMostPoints() into fAverageTrack().
- Added description and examples to several large or important functions  

------------
# Version:
20120115

# Changes:
- Create two functions to read in *.gpx files. 
- Function read1GPX() reads one file and converts it to a data.frame or to a SpatialPointDataFrame
- Function readGPX() reads a vector of filenames and calls for each filename read1GPX() resulting in a list.
- Added several lines to write output to a logfile. 
- Changed several defaults. I only want to average the input. Input is therefore limited to 2 files.

------------
# Version:
20120103

# Changes:
- Create several overal functions to organise the scripts and flow of the program

------------
# Version:
20111230

# Changes:
- wrote functions to find closest point on a known line from a given point with slope and intercept (see "findIntersectionWithorthogonalLine.R")
- put code to find track with most points in its own function (fFindTrackWithMostPoints(track1, track2))
- Wrote a function to average two lines based on closest point on line to point.
- Wrote a wrapper to choose which of the methods you want to use for averaging. Moth methods are also possible.
- Wrote a function fCreateTrack() to create routes from two lists of points. 

# TO DO
- When averaging 'lijn1' and 'lijn2' from test grootLeggelderveld, some points fail with method "l". 
  See fore example point 7, 10, 11, (13, 14, 15?) and 16. Test to check if point is on line fails.

------------
# Version:
20111213

# Changes:
- Rewrote loadFiles.R to be a function. Add more error handling.
- Rewrote findIntersection.R to be a function. Add more error handling.
- Rewrote NewTracks.R to be a function.
- Moved fAverageTrack() from functions.R to its one file fAverageTrack.R
- Wrote two functions. They sort the track so the smallest longtitude of begin or endpoint is set as first point(x). 
	Thus sorting from west to east. When both points hava the same longtidude sorting is done on lattitude (y). Sorting from south to north.
- Added a function 'fSaveToOneTrack()' to save all tracks into one gpx file with several track segements (trkseg).

# TO DO
- Averaging tracks and paths and clearing spike doesn't work 100% correct. 
- Probably, i have to write more in functions, so i can re-use stuff more. 
- Search for paralel tracks to average

------------
# Version:
20111203

# Changes:
- Changed the working of rounding the numbers. R always calculates with full precision, even when less digits are shown. Rounding is not needed, only for writing of results. Export of the intersections and writing the gpx-files now uses global variable 'roundTo'. All other roundings are removed.
- Cleaned up all the code
- Added a second intersection if a route intersects with itself. Removed again because in newTracks I uncommented a test that both routeID's of a intersections couldn't be the same. But ofcourse they can. 
- Added code to search for tracks to average
- Added code to search for tracks with same begin and end point. 

# TO DO
- Averaging tracks and paths and clearing spike doesn't work 100% correct. 
- Probably, i have to write more in functions, so i can re-use stuff more. 
- Search for paralel tracks to average

------------
# Version:
20111130

# Changes:
- Moved the code to write a gpx-track into fSaveTrack() and moved to functions.R. 
- Cleaning up of fCalculateSlope()
- In loadFiles.R use function fCalculateSlope() to calculate slopes for each Route added.
- In NewTracks.R, calculate slope with fCalculateSlope() and write gpx-files with fSaveTracks()
- In NewTracks.R added test to see if dd (all intersections for a track) actually has rows (intersections). If so, continue
- Created a function to create a point to be added; fPointToAdd().
- In NewTracks.R used the function fPointToAdd().
- Tried to make some sense out of the averageLine() function and the distanceToLine() function. The seem to work again. 
- busy adding code to search for 2 lines with the same begin and endpoint. They should be averaged.

# TO DO
- fix the resulting tracks. It's actually done. There is still a minor detail to work out. If a intersection is the begin and end point of a track, the endpoint is not found...
- Continue with cleaning up the code in other files.
- Search for round trips and spikes
- Search for paralel tracks to average

------------
# Version:
20111123

# Changes:
- Setting up projectdir and testfiles.
- Changed call to columns from dd[,x] to dd$x[] to make changing the format easier.
- Added the possibilty to write *.gpx files from tracks
- updated the code on readabilty and removed obsolete code.
- Worked up to "loadFiles.R", "findIntersections.R", "newTracks.R" and "saveTracks.R". Program runs, but the resulting tracks are not correct.
- Added garbage collection

# TO DO
- fix the resulting tracks. It's actually done. There is still a minor detail to work out. If a intersection is the begin and end point of a track, the endpoint is not found...
- Continue with cleaning up the code in other files.
- Search for round trips and spikes
- Search for paralel tracks to average




average-tracks.R
# average tracks is an R script to make an "average" track of several GPS
# tracks by combining/merging them into a "line of best fit". Made by
# Michiel Faber <http://www.openstreetmap.org/user/Michiel%20Faber>
# This software is dedicated to the public domain through CC0:
# http://creativecommons.org/publicdomain/zero/1.0/
#
# See http://wiki.openstreetmap.org/wiki/average_tracks for usage instructions.
# Home: https://github.com/dittaeva/average-tracks
# version: 0.4
# date: 31-aug-2011

# Save the start time
beginTime <- proc.time()


# Import all *.csv files found in the working dir
listOfFiles <- list.files(pattern = ".csv")
fileCount <- length(listOfFiles)
ntracks <- 0
for (n in 1:fileCount) {
# read files
  fileName <- listOfFiles[n]
  route <- read.csv(file=fileName, header=FALSE)  # dont import the header

# test if route contains points. We could have clipped the complete route 
  if (nrow(route) > 0) {
# overwrite the inputfile with the clipped route
#    write.table(route, file = fileName, quote = FALSE, row.names=FALSE, col.names=FALSE, sep=",", dec=".")

    ntracks <- ntracks + 1 # count the number of routes
    colnames(route) <- c("lon", "lat", "routeID") # set column names
    route[,3] <- rep(ntracks,each=nrow(route)) # each route gets it own routeID
#combine the routes
    if (ntracks == 1) { 
      tracks <- route # the first route is added directly
    } else {
      tracks <- rbind(tracks,route) # all later routes are appended
    }
  }
}

# Find the track with the highest number of points
# count per id
# find the highest count
max_points=0
for (i in 1:ntracks) { # search in all routeID's
  tmpdf <- tracks[tracks$routeID==i, ] # [,3] is the column with the routeID
  if (max_points < nrow(tmpdf)) {
    max_points <- nrow(tmpdf)
    refRouteID <- i # refRoute is the route with the moist gps-points.
    refRouteID
  }
}
refRouteID

# Find distance between two points
refRouteXcor <- tracks[tracks$routeID==refRouteID, 1]
refRouteYcor <- tracks[tracks$routeID==refRouteID, 2]
newTracks <- tracks[tracks$routeID==refRouteID, ] # store refRoute in a new list of routes
averageTrack <- newTracks
averageTrack$lon <- 0 # set all x coordinates to 0
averageTrack$lat <- 0 # set all y coordinates to 0
averageTrack <- averageTrack[,-3] # remove the routeID
#averageTrack[, 3] <- ntracks+1 # set correct routeID
 
# for all points in refRoute -> k
# for all tracks in tracks -> l
# for all points in one track -> m
for (l in 1:ntracks) {
  if (l != refRouteID) {
    for (k in 1:max_points) {
      RouteXcor <- tracks[tracks$routeID==l, 1]
      RouteYcor <- tracks[tracks$routeID==l, 2]
      shortest_distance=999999
      for (m in 1:length(RouteXcor)) {
        distance <- sqrt((RouteXcor[m]-refRouteXcor[k])^2+(RouteYcor[m]-refRouteYcor[k])^2)
        if (distance < shortest_distance) {
          shortest_distance <- distance
          pointID <- m
        }
      }
      newTracks <- rbind(newTracks, c(RouteXcor[pointID], RouteYcor[pointID], l)) # add closest point to refRoute-point for each route
    }
  }
}

# Average all tracks into one single track
for (n in 1:(max_points)) {
  for (m in 1:(ntracks)) {
    if (m == 1) {
      valueX <- newTracks[n, 1]
      valueY <- newTracks[n, 2]
    } else {
      valueX <- valueX + newTracks[n+(m-1)*max_points, 1]
      valueY <- valueY + newTracks[n+(m-1)*max_points, 2]
    }
  }
  averageTrack[n, 1] <- valueX / ntracks
  averageTrack[n, 2] <- valueY / ntracks
}

# cleaning up
#averageTrack <- averageTrack[,-3] # remove the routeID
averageTrack <- round(averageTrack, 6) # round to 6 decimal places

cat("Done. \n")
cat("The dataframe 'averageTrack' contains the newly computed average track. \n")
cat("\n")

#Saving the resulting track in averageTrack.csv.
cat("Saving the resulting track in averageTrack.csv. \n")
write.table(averageTrack, file = "averageTrack.csv", quote = FALSE, row.names=FALSE, col.names=FALSE, sep=",", dec=".")

# Show duration of script
endTime <- proc.time()
duration <- endTime - beginTime
duration



programAverage.R
# version: 12.01.21
# creator: Michiel Faber
# longtitude: east-west; x
# lattitude: north-south; y
rm(list=ls())

# Working directory. Directory where programAverage.R and functions are located. Also where relative paths relate to. Uncomment and customise if needed, or current directory will be used.
#setwd("~/bin/average-tracks/") 

source("functionsGlobal.R")
source("functionsAverage.R")

loadedTracks <- fLoadFiles(
	inputDir="input", # Directory with input files.
	listOfFiles=NULL,
	fileType="gpx", # "gpx" or "csv".
	nTracks=0,
	loadHeader=FALSE, # csv files without a header.
	method="rgdal") # "rgdal" or "maptools". Both will autoinstall if necessary libraries are present.

fileNames <- loadedTracks[[2]]
tracks <- loadedTracks[[1]]

averageTrack <- fAverageTrack(
	tracks[tracks$routeID==1,], # track 1
	tracks[tracks$routeID==2,], # track 2
	elevation = TRUE,
	method="b") # Combine both averaging methods. "p" for only point method, "l" for line method.

# Uncomment png() to save plot as png, also uncomment dev.off() on line 42.
#png(filename = "output/averageTrack.png", width=2048, height=2048, res=200)
 plot(tracks[tracks$routeID==1,2:1], type="l", pch=2, col="green")
 lines(tracks[tracks$routeID==2,2:1], type="l", pch=2, col="green")
 lines(averageTrack[,2:1], type="l", pch=1, col="red")

 legend.text <- c(fileNames[1], fileNames[2], "Average")
 legend.col <- c("green", "blue", "black")
 legend.lty <- "solid"
 legend.pch <- c(2,1,4) 
 legend.position <- "topleft"
 legend(x=legend.position, legend=legend.text, col=legend.col, lty=legend.lty, pch=legend.pch)
#dev.off() # Uncomment to save plot.

fSaveTrack(
	track=averageTrack,
	outputDir="output", # Directory where output file will be saved.
	trk="20120121", # Part of the trackname. "track" will be placed in front of "trk" value.
	roundTo=6)



functionsGlobal.R
#-----------------------------


###########################
#   function fLoadFiles   #
###########################
fLoadFiles <- function(inputDir, listOfFiles=NULL, fileType="csv", nTracks=0, loadHeader=FALSE, method="maptools") {

# Description:
# 	Import set of points (representing lines) stored in *.gpx, *.csv files or user specified files from a given directory.
# 	Additional to loading, it sorts the points and add line characteristics to each point.
#
# Usage:
#
#	fLoadFiles(inputDir, listOfFiles=NULL, fileType="csv", nTracks=0, loadHeader=FALSE, method="maptools")
#
# Arguments:
#
# 	      inputDir:	Directory where the files to import are in.
#
#	   listOfFiles:	A vector of files to import.
#
# 	      fileType:	type of files to imported. Currently only "gpx" and "csv" are supported. Only used when listOfFiles is not given.
#
# 	       nTracks:	number to start identifying each set of points. If you want to add to existing set of points increase this value appropiatly.
#
# 	    loadHeader:	logical value indicating if a header is present in the csv's files. 
#
# 	        method:	method to use when importing *.gpx files. Must be "maptools" or "rgdal". Both methods use the load functions from their 
#			respective packages.
#
# Details:
#
#	... ... ...
#
# Value:
#
#	A list with on first position a data frame with all routes and their points. Second is a list of the imported files.
#
# Author(s):
#
#	Michiel Faber
#
#
# References:
#
#	... ... ...
#
# See Also:
#
#	... ... ...
#

 columnNames <- c("lat", "lon", "height", "routeID", "pointID", "slope", "intercept", "pointType", "nJoins")

 if (nTracks == 0) {
  Tracks <- NULL # Define an empty Tracks object
 }
 if (is.null(listOfFiles)) {
  listOfFiles <- list.files(inputDir, pattern = fileType) 
 }
 fileCount <- length(listOfFiles) # number of files

 if (fileCount > 0) {
  for (fileNumber in 1:fileCount) {
  # read files
   fileName <- listOfFiles[fileNumber]
  if (fileType=="csv") {
   Route <- read.csv(file=paste(inputDir,fileName,sep="/"), header=loadHeader)  # dont import a header, because there is none
  }
  if (fileType=="gpx") {
   tmpRoute <- read1GPX(paste(inputDir,fileName,sep="/"), methodToUse=method)
   if (method=="maptools") {
    Route <- as.data.frame(tmpRoute[,3:4])
    Route[,3] <- as.numeric(substring(as.character(tmpRoute[,14]), 1, (nchar(as.character(tmpRoute[,14]))-1)))
   }
   if (method=="rgdal") {
    Route <- as.data.frame(tmpRoute@coords[,2:1])
    Route[,3] <- as.data.frame(tmpRoute$ele)
   }
  }
   if (nrow(Route) > 0) {  # test if route contains points.
    nTracks <- nTracks + 1 # keep count of the number of routes
    Route[,4] <- nTracks  # each route gets it own routeID
    Route[,5] <- c(1:nrow(Route)) # each point gets a pointID
    Route[,6] <- NA 
    Route[,7] <- NA
    Route[,8] <- NA  
    Route[,9] <- NA
    colnames(Route) <- columnNames # set column names
    Route <- fSortOnDirection(Route)
    Route <- fCalculateSlope(Route)
 #   Route <- fSetPointTypes(Route)
    # combine the routes
    if (nTracks == 1) { 
     Tracks <- Route # the first route is added directly
    } else {
     Tracks <- rbind(Tracks,Route) # all later routes are appended
    }
   }
  }
 }
 result <- list(Tracks, listOfFiles)
return(result)
}

read1GPX <- function(x, methodToUse="maptools") {
# http://stackoverflow.com/questions/6397523/read-multiple-gpx-files
 if (methodToUse=="maptools") {
  if(!require("maptools")){
   install.packages("maptools")
   if(!require("maptools")){
    stop("could not install pacakge 'maptools'.")
   }
  }
  dd <- readGPS(i = "gpx", f = x, type = "w")
 }
 if (methodToUse=="rgdal") {
  if(!require("rgdal")){
   install.packages("rgdal")
   if(!require("rgdal")){
    stop("could not install package 'rgdal'.")
   }
  }
  dd <- readOGR(x, layer="track_points")
 }
return(dd)
}

readGPX <- function(listOfFiles=NULL, method="maptools") {
# http://stackoverflow.com/questions/6397523/read-multiple-gpx-files
 if (is.null(listOfFiles)) {
  out <- NULL
  stop("Please provide a vector with filenames")
 }
 if (method=="maptools") {
  out <- lapply(listOfFiles, FUN=read1GPX)
 }
 if (method=="rgdal") {
  out <- lapply(listOfFiles, FUN=read1GPX, methodToUse=method)
 }
 names(out) <- listOfFiles
return(out)
}

fSortOnDirection <- function(track) {
 nPoints <- nrow(track)
 if (track$lon[1] == track$lon[nPoints]) {
  orderValues <- fOrderPoints(track$lat)
 } else {
  orderValues <- fOrderPoints(track$lon)
 }
 track <- track[orderValues,]
 track$pointID <- c(1:nPoints)
return(track)
}

fOrderPoints <- function(values) {
 nPoints <- length(values)
 orderValues <- c(1:nPoints)
 maxValue <- max(values[1], values[nPoints])
 selectMaxPoint  <- maxValue == c(values[1], values[nPoints])
 if (selectMaxPoint[1]) {
  orderValues <- c(nPoints:1)
 }
return(orderValues)
}

#-----------------------------


################################
#   function fCalculateSlope   #
################################

fCalculateSlope <- function(track) {
 # ============
# Description:
#
# 	This function calculates for a set of points (representing a line), for each segment between the points the intersect a slope 
#	of the line coming from the previos point. Therefor point one has no intercept or slope.
#
# Usage:

# 	fCalculateSlope(track)
#
# Arguments:
# 		 track: a data frame of coordinates. Column names should be provided and should include "lon" and "lat". 
#
# Details:
#
#	... ... ...
#
# Value:
#
#	A data frame with the original columns and 2 extra colums with names "slope" and "intercept".
#
# Author(s):
#
#	Michiel Faber
#
#
# References:
#
#	... ... ...
#
# See Also:
#
#	'fCreateTrack.
#
#
# Examples:
#	... ... ...

 comment <- TRUE # only dummy to prevent executing of example
 if (!comment) {

  lat <- c(1,2,3.5,4,5)
  lon <- c(2,3,4,5,6)
  track <- data.frame(lat,lon)
  newTrack <- fCalculateSlope(track)
  plot(newTrack[,2:1],type="b", xlim=c(-1,7),ylim=c(-3,7))
  for (x in 2:5) {
   lines(c(-1:7),newTrack$slope[x]*c(-1:7)+newTrack$intercept[x],col="green")
  }
  lines(newTrack[,2:1],type="b", xlim=c(-1,7),ylim=c(-3,7))

 } # end comment

######
# Begin function
######

 # Define function to be used by sapply()
 f <- function(i) { 
  if ((track$lon[i]-track$lon[i-1]) == 0) { # if line goes vertical...
   # to do: give a and b better values
   slope <- 999999
   intercept <- 999999
   result <- cbind(slope,intercept)
  } else {
   slope <- (track$lat[i]-track$lat[i-1]) / (track$lon[i]-track$lon[i-1])
   intercept <- track$lat[i-1] - slope * track$lon[i-1]
   result <- cbind(slope,intercept)
  }
 return(result)
 }
 slopes <- sapply(2:nrow(track), FUN=f)
 track$slope[2:nrow(track)] <- t(slopes)[,1]
 track$intercept[2:nrow(track)] <- t(slopes)[,2]
return(track)
}

fSetPointTypes <- function(track, beginType="b", midType="l", endType="e") {
 track$pointType[1] <- NA
 track$pointType[2:(nrow(track)-1)] <- midType # This one come first to make sure first and last point get the good values when nrow(track) = 2
 track$pointType[1] <- beginType
 track$pointType[nrow(track)] <- endType
return(track)
}

fSaveTrack <- function(track, outputDir, trk=NULL, roundTo=6) {
# function to save a track to GPX.
# If everything is converted to Spatial-classes rgdal/ogr functions can be used.
# Define gpx header
 lineHeader1 <- "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\" ?>" 
 lineHeader2 <- "<gpx xmlns=\"http://www.topografix.com/GPX/1/1\" version=\"1.1\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:schemaLocation=\"http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd\" creator=\"M. Faber\">"
 lineTrkStart <- "<trk>"
 lineTrkSegStart <- "<trkseg>"
 # lineTrkptStart <- # see below
 lineTrkptEnd <- "</trkpt>"
 lineTrkSegEnd <- "</trkseg>"
 lineTrkEnd <- "</trk>"
 lineFooter1 <- "</gpx>"

 if (is.null(trk)) {
  trk <- track$routeID[1]
 }
 trackName <- paste("Track:", trk, sep=" ")
 trackDescription <- "These are my first tracks created by R"
 lineTrackName <- paste("<name>", trackName, "</name>", sep="")
 lineTrackDescription <- paste("<desc>",  trackDescription, "</desc>", sep="")
 fileName <- paste("track", trk, ".gpx", sep="")
 outputName <- paste(outputDir, fileName, sep="/")
 cat(paste(lineHeader1, lineHeader2, lineTrkStart, lineTrackName, lineTrackDescription, lineTrkSegStart, "", sep="\n"), file=outputName)
 for (pnt in unique(track$pointID)) { 
  lineTrkptStart <- paste("<trkpt lon=\"", round(track$lon[pnt],roundTo), "\" lat=\"", round(track$lat[pnt],roundTo), "\">", sep="")
  if (!is.null(track$height[pnt])) {
   lineEle <- paste("<ele>", round(track$height[pnt],roundTo), "</ele>", sep="")
  } else { 
   lineEle <- paste("<ele>", NA, "</ele>", sep="") 
  }
  cat(paste(lineTrkptStart, lineEle, lineTrkptEnd, "", sep="\n"), file=outputName,  append=T)
 }
 cat(paste(lineTrkSegEnd, lineTrkEnd, lineFooter1, "", sep="\n"), file=outputName, append=T)
}

fSaveToOneTrack <- function(newTracks, trackName="name", outputDir=outputDir) {
# Define gpx header
 lineHeader1 <- "<?xml version=\"1.0\" encoding=\"UTF-8\" standalone=\"no\" ?>" 
 lineHeader2 <- "<gpx xmlns=\"http://www.topografix.com/GPX/1/1\" version=\"1.1\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xsi:schemaLocation=\"http://www.topografix.com/GPX/1/1 http://www.topografix.com/GPX/1/1/gpx.xsd\" creator=\"M. Faber\">"
 lineTrkStart <- "<trk>"
 lineTrkSegStart <- "<trkseg>"
 # lineTrkptStart <- # see below
 lineTrkptEnd <- "</trkpt>"
 lineTrkSegEnd <- "</trkseg>"
 lineTrkEnd <- "</trk>"
 lineFooter1 <- "</gpx>"

 trackDescription <- "These are my first tracks created by R"
 lineTrackName <- paste("<name>", trackName, "</name>", sep="")
 lineTrackDescription <- paste("<desc>",  trackDescription, "</desc>", sep="")
 fileName <- paste(trackName, ".gpx", sep="")
 outputName <- paste(outputDir, fileName, sep="/")
 cat(paste(lineHeader1, lineHeader2, lineTrkStart, lineTrackName, lineTrackDescription, "", sep="\n"), file=outputName)
 for (trkID in unique(newTracks$routeID)) {
  selectTrack <- newTracks$routeID==trkID
  workTrack <- newTracks[selectTrack,]
  # function to save a track to GPX.
  cat(paste(lineTrkSegStart, "", sep="\n"), file=outputName,  append=T)
  for (pnt in unique(workTrack$pointID)) { 
   lineTrkptStart <- paste("<trkpt lon=\"", round(workTrack$lon[pnt],roundTo), "\" lat=\"", round(workTrack$lat[pnt],roundTo), "\">", sep="")
   if (!is.null(workTrack$height[pnt])) {
    lineEle <- paste("<ele>", round(workTrack$height[pnt],roundTo), "</ele>", sep="")
   } else { 
    lineEle <- paste("<ele>", NA, "</ele>", sep="") 
   }
   cat(paste(lineTrkptStart, lineEle, lineTrkptEnd, "", sep="\n"), file=outputName,  append=T)
  }
  cat(paste(lineTrkSegEnd, "", sep="\n"), file=outputName, append=T)
 }
 cat(paste(lineTrkEnd, lineFooter1, "", sep="\n"), file=outputName, append=T)
}

